hdfs3/tests/test_hdfs3.py

```diff
@@ -456,3 +503,133 @@ def test_different_handles_in_processes():
 
     for p in procs:
         p.join()
+
+
+def test_ensure():
+    assert isinstance(ensure_bytes(''), bytes)
+    assert isinstance(ensure_bytes(b''), bytes)
+    assert isinstance(ensure_string(''), unicode)
+    assert isinstance(ensure_string(b''), unicode)
+    assert ensure_string({'x': b'', 'y': ''}) == {'x': '', 'y': ''}
+    assert ensure_bytes({'x': b'', 'y': ''}) == {'x': b'', 'y': b''}
+
+
+
+def test_touch_exists(hdfs):
+    hdfs.touch(a)
+    assert hdfs.exists(a)
+
+
+def test_write_in_read_mode(hdfs):
+    hdfs.touch(a)
+
+    with hdfs.open(a, 'r') as f:
+        with pytest.raises(IOError):
+            f.write(b'')
+
+
+def test_readlines(hdfs):
+    with hdfs.open(a, 'w') as f:
+        f.write(b'123\n456')
+
+    with hdfs.open(a, 'r') as f:
+        lines = f.readlines()
+        assert lines == [b'123', b'456']
+
+    with hdfs.open(a, 'w') as f:
+        with pytest.raises(IOError):
+            f.read()
+
+
+def test_put(hdfs):
+    data = b'1234567890' * 10000
+    with tmpfile() as fn:
+        with open(fn, 'wb') as f:
+            f.write(data)
+
+        hdfs.put(fn, a)
+
+        assert hdfs.cat(a) == data
+
+
+def test_getmerge(hdfs):
+    with hdfs.open(a, 'w') as f:
+        f.write(b'123')
+    with hdfs.open(b, 'w') as f:
+        f.write(b'456')
+
+    with tmpfile() as fn:
+        hdfs.getmerge('/tmp/test', fn)
+
+        with open(fn, 'rb') as f:
+            data = f.read()
+        assert data == b'123456'
+
+
+def test_get(hdfs):
+    data = b'1234567890'
+    with tmpfile() as fn:
+        with hdfs.open(a, 'w') as f:
+            f.write(data)
+
+        hdfs.get(a, fn)
+
+        with open(fn, 'rb') as f:
+            data2 = f.read()
+        assert data2 == data
+
+    with pytest.raises(IOError):
+        hdfs.get(b, fn)
+
+
+def test_open_errors(hdfs):
+    hdfs.touch(a)
+    with pytest.raises(ValueError):
+        hdfs.open(a, 'r', block_size=1000)
+
+    hdfs.disconnect()
+    with pytest.raises(IOError):
+        hdfs.open(a, 'w')
+
+
+def test_du(hdfs):
+    with hdfs.open(a, 'w') as f:
+        f.write(b'123')
+    with hdfs.open(b, 'w') as f:
+        f.write(b'4567')
+
+    assert hdfs.du('/tmp/test') == {a: 3, b: 4}
+    assert hdfs.du('/tmp/test/', total=True) == {'/tmp/test/': 3 + 4}
+
+
+def test_get_block_locations(hdfs):
+    with hdfs.open(a, 'w') as f:
+        f.write(b'123')
+
+    locs = hdfs.get_block_locations(a)
+    assert len(locs) == 1
+    assert locs[0]['length'] == 3
+
+
+def test_chmod(hdfs):
+    hdfs.touch(a)
+    hdfs.chmod(a, 'r')
+
+    with pytest.raises(IOError):
+        hdfs.open(a, 'a')
+
+
+@pytest.mark.xfail
+def test_chmod_write(hdfs):
+    hdfs.chmod(a, 'r')
+    hdfs.chmod(a, 'w')
+
+    with hdfs.open(a, 'a') as f:
+        pass
+
+
+@pytest.mark.xfail
+def test_chown(hdfs):
+    hdfs.touch(a)
+    i = hdfs.info(a)
```

Comment:
I'm not sure what's going on in the xfailed tests.  @martindurant were you ever able to get this functionality to work?